### PR TITLE
Llama2 embedding

### PR DESCRIPTION
**Pull request by nigel-daniels**
I've added an embedding module based on the `node-llama-cpp` codebase. This now gives developers a full range of Llama2 modules to work with a locally running Llama2 LLM. 

With this module I have also refactored how the Llama2 model, context and sessions get created to ensure this is consistent across the LLM, ChatLLM and Embedding modules.

### CODE REVIEW

File: langchain/src/chat_models/llama_cpp.ts [hunk 1]
```diff
@@ -99,52 +61,39 @@ export class ChatLlamaCpp extends SimpleChatModel<LlamaCppCallOptions> {
...
+      topP: this.topP,
+      trimWhitespaceSuffix: this.trimWhitespaceSuffix,
+    };
+  }
+
   /** @ignore */
   async _call(
     messages: BaseMessage[],
+    // @ts-expect-error - TS6133: 'options' is declared but its value is never read.
     options: this["ParsedCallOptions"]
```
Comment (jacoblee93) [inline langchain/src/chat_models/llama_cpp.ts:97]: Just use `_options`

File: langchain/src/chat_models/llama_cpp.ts [hunk 2]
```diff
@@ -99,52 +61,39 @@ export class ChatLlamaCpp extends SimpleChatModel<LlamaCppCallOptions> {
...
-    this.logitsAll = inputs?.logitsAll;
-    this.lowVram = inputs?.lowVram;
-    this.modelPath = inputs.modelPath;
-    this.seed = inputs?.seed;
-    this.useMlock = inputs?.useMlock;
-    this.useMmap = inputs?.useMmap;
-    this.vocabOnly = inputs?.vocabOnly;
-    this._model = new LlamaModel(inputs);
-    this._context = new LlamaContext({ model: this._model });
+    this.maxTokens = inputs?.maxTokens;
```
Comment (jacoblee93) [inline langchain/src/chat_models/llama_cpp.ts:64]: Why remove so many parameters? Seems out of scope for this PR?

File: docs/docs/modules/data_connection/text_embedding/integrations/llama_cpp.mdx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+npm install -S node-llama-cpp
+```
+
+You will also need a local Llama 2 model (or a model supported by [node-llama-cpp](https://github.com/withcatai/node-llama-cpp)). You will need to pass the path to this model to the LlamaCpp module as a part of the parameters (see example).
+
+Out-of-the-box `node-llama-cpp` is tuned for running on a MacOS platform with support for the Metal GPU of Apple M-series of processors. If you need to turn this off or need support for the CUDA architecture then refer to the documentation at [node-llama-cpp](https://withcatai.github.io/node-llama-cpp/).
+
+For advice on getting and preparing `llama2` see the documentation for the LLM version of this module.
+
+A note to langchainjs contributers: if you want to run the tests associated with this module you will need to put the path to your local model in the environment variable `LLAMA_PATH`.
```
Comment (jacoblee93) [inline docs/docs/modules/data_connection/text_embedding/integrations/llama_cpp.mdx:27]: `LangChain.js contributors`

File: langchain/src/embeddings/llama_cpp.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    });
+  }
+
+  /**
+   * Generates an embedding for a single text.
+   * @param text - A string to generate an embedding for.
+   * @returns A Promise that resolves to an array of numbers representing the embedding.
+   */
+  async embedQuery(text: string): Promise<number[]> {
+    return new Promise((resolve) => {
```
Comment (jacoblee93) [inline langchain/src/embeddings/llama_cpp.ts:63]: Why the wrapper here?

File: langchain/src/embeddings/llama_cpp.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      const tokens: number[] = [];
+
+      const encodings = this._context.encode(text);
+      for (let i = 0; i < encodings.length; i += 1) {
+        const token: number = +encodings[i];
```
Comment (jacoblee93) [inline langchain/src/embeddings/llama_cpp.ts:68]: Is the `+` cast necessary?

File: langchain/src/embeddings/llama_cpp.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    return new Promise((resolve) => {
+      const tokensArray = texts.map((text) => this._context.encode(text));
+
+      const embeddings: number[][] = [];
+
+      for (const tokens of tokensArray) {
+        const embedArray: number[] = [];
+
+        for (let i = 0; i < tokens.length; i += 1) {
+          const nToken: number = +tokens[i];
```
Comment (jacoblee93) [inline langchain/src/embeddings/llama_cpp.ts:46]: Is the `+` cast necessary?

**Review by jacoblee93 — COMMENTED** [review]
Thanks for this! A few questions

File: langchain/src/embeddings/llama_cpp.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    this._context = createLlamaContext(this._model, _inputs);
+  }
+
+  /**
+   * Generates embeddings for an array of texts.
+   * @param texts - An array of strings to generate embeddings for.
+   * @returns A Promise that resolves to an array of embeddings.
+   */
+  async embedDocuments(texts: string[]): Promise<number[][]> {
+    return new Promise((resolve) => {
```
Comment (jacoblee93) [inline langchain/src/embeddings/llama_cpp.ts:37]: We can also take advantage of concurrent calls and retries if we use `this.caller.call` (standard property on `Embeddings`

https://github.com/langchain-ai/langchainjs/blob/main/langchain/src/embeddings/openai.ts#L246

File: langchain/src/embeddings/llama_cpp.ts (hunk 5 above)
Comment (nigel-daniels) [reply langchain/src/embeddings/llama_cpp.ts:68]: The `node-llama-cpp` call responds with something that looks like:
```
{"1": "1234", "2": "5678", "3": "90" }
```
As a result the values are `string` types, the `+` is to coerces them to be `number`.

File: langchain/src/embeddings/llama_cpp.ts (hunk 6 above)
Comment (nigel-daniels) [reply langchain/src/embeddings/llama_cpp.ts:46]: See previous comment.

File: langchain/src/embeddings/llama_cpp.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (nigel-daniels) [reply langchain/src/embeddings/llama_cpp.ts:37]: I need to return a promise and the call to the `encode` function is a straight up synchronous call.

File: langchain/src/embeddings/llama_cpp.ts (hunk 4 above)
Comment (nigel-daniels) [reply langchain/src/embeddings/llama_cpp.ts:63]: See previous comment.

File: langchain/src/embeddings/llama_cpp.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (nigel-daniels) [reply langchain/src/embeddings/llama_cpp.ts]: Using the async caller I can drop the wrapper.

[the remark this answers is not in the document]
Comment (jacoblee93) [reply langchain/src/embeddings/llama_cpp.ts]: Sounds good - you can return sync values in an `async` function though as they'll be automatically wrapped!

**Comment by jacoblee93** [comment]
LGTM, thanks!

**Comment by jacoblee93** [comment]
Hey @nigel-daniels thank you so much for all this - it's going to be really helpful to a lot of people! 

Separately, would love to know if you're planning on building anything with this yourself?

**Comment by nigel-daniels** [comment]
Hi @jacoblee93 my pleasure, I hope everyone get's a lot of use out of this. Happy to chat about my plans for these tools maybe you can DM me on Discord (JSDevGuy).
